platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java

```diff
@@ -0,0 +1,208 @@
+            decryptedDictionaryToReturn.put(key, decryptIfNecessary(key, dict.get(key)));
+        }
+        return decryptedDictionaryToReturn;
+    }
+
+    private Object encryptIfNecessary(String key, Object value) {
+        if (value instanceof String && !EXCLUDED_PROPERTIES.contains(key)) {
+            return agent.encrypt((String) value);
+        }
+        return value;
+    }
+
+    private Object decryptIfNecessary(String key, Object value) {
+        if (value instanceof String && !EXCLUDED_PROPERTIES.contains(key)) {
+            return agent.decrypt((String) value);
+        }
+        return value;
+    }
+
+    /**
+     * PAX Logging might not be fully configured and ready if there is a failure in the agent's
+     * constructor. To remedy this, we will cache failure data to later throw back.
+     */
+    private class EncryptionAgent {
+        private String initFailureMessage = null;
+
+        private Exception initException = null;
+
+        private Crypter crypter;
+
+        EncryptionAgent(String pwDirString) {
+            File pwDir = new File(pwDirString);
+            if (!pwDir.mkdirs() && !pwDir.exists()) {
+                initFailureMessage = "Cannot create required directory for encryption ["
+                        + pwDir.getAbsolutePath() + "]. This might be a permissions issue. Check "
+                        + "OS settings, file permissions, and the active user account. Then unzip a "
+                        + "new instance and try again";
+                throw new IllegalStateException(initFailureMessage);
+            }
+
+            String[] children = pwDir.list();
+            if (children == null) {
+                initFailureMessage = "Expected directory [" + pwDir.getAbsolutePath() + "] to "
+                        + "exist and be accessible. Check OS settings, file permissions, and the "
+                        + "active user account. Then unzip a new instance and try again";
+                throw new IllegalStateException(initFailureMessage);
+            }
+
+            synchronized (EncryptionAgent.class) {
+                if (!Arrays.asList(children)
+                        .contains("meta")) {
+                    KeyczarTool.main(new String[] {"create", "--location=" + pwDirString,
+                            "--purpose=crypt", "--name=Password"});
+                    KeyczarTool.main(new String[] {"addkey", "--location=" + pwDirString,
+                            "--status=primary"});
+                }
+                try {
+                    this.crypter = new Crypter(pwDirString);
+                } catch (Exception e) {
+                    initFailureMessage = "Could not setup encryption. ";
+                    initException = e;
+                    throw new IllegalStateException(initFailureMessage, initException);
+                }
+            }
+        }
+
+        synchronized String encrypt(String plainTextValue) {
```

Comment:
❓ Why are those methods `synchronized`?

Reply:
Took this straight from the encryption service so I'm not 100% certain. Upon further investigation, I'm inclined to agree with you: 
https://github.com/google/keyczar/blob/master/java/code/src/org/keyczar/KeyczarFileReader.java#L62-L75
It appears that Crypters, which extend Encrypters, which extend Keyczar, will cache everything they need: 
https://github.com/google/keyczar/blob/master/java/code/src/org/keyczar/Keyczar.java#L79-L108